techmatters/serverless
# CSAM IWF reporting from Aselo: API integration

Jira: https://bugs.benetech.org/browse/CHI-704

This PR is related to:
- https://github.com/techmatters/flex-plugins/pull/543
- https://github.com/techmatters/hrm/pull/131

This PR adds a new function (`reportToIWF`) to submit reports to IWF from our system.
It expects the following inputs:
- `Reported_URL` **required**: a string representing the url to the content being reported.
- `Reporter_Anonymous` **required**: either Y or N to indicate if the report is anonymous or not
- `Reporter_First_Name`: reporter first name (used if not anonymous).
- `Reporter_Last_Name` reporter last name (used if not anonymous).
- `Reporter_Email_ID`: reporter email address (**required** if not anonymous).
- `Reporter_Description` report brief description.

~WIP:~
- ~Unit tests.~
- ~Move the function to be protected before merging (so it's not public).~
- ~Include IWF credentials on deploy workflow.~

To test:
- Checkout this branch.
- Create a `.env` file in the root folder and fill this variables 
  - IWF_API_USERNAME
  - IWF_API_PASSWORD
  - IWF_API_URL
 
  (look for the values on Parameter Store using the same names).
- Run `npm ci` to install deps and then `npm start` to start local server.
- Hit http://localhost:3000/reportToIWF?Reported_URL=https://something.com&Reporter_Anonymous=Y in your browser or using curl
  ```
  curl -X POST -H 'Content-type: application/json' --data '{"Reported_URL": "something", "Reporter_Anonymous": "Y"}' http://localhost:3000/reportToIWF
  ```
- Play around using different values for the parameters.

What to expect: the function only rejects the request prior to hit IWF API if the **required** parameters are missing or invalid, but sends it otherwise, returning the same response that IWF sends. This means that, if for example, we submit a non-anonymous report without `Reporter_Email_ID`, the function will hit IWF but this will be rejected. We can validate this for ourselves but IWF is already doing some work on the inputs, like checking that email is a valid address. If we go with this approach, we rely on IWF validations, saving ourselves from this work.
Example response when sending a non anonymous report with a non empty but invalid email address (asdasd@asdasd).
![Screenshot from 2021-11-10 14-24-38](https://user-images.githubusercontent.com/15805319/141162393-9f2ae11e-7195-4115-97e4-be54fc372db8.png)
Each input has it's own limitations and IWF performs the pertinent checks, so I don't think we should duplicate all of those, but let me know if you think different!
